### DISCUSSION TITLE
create a solid out of a shell

### DISCUSSION THREAD

**Discussion by raydeleu** [donalffons/opencascade.js#142] [Q&A]
I am using opencascade.js as a part of CascadeStudio. I ran into the following problem that I also posted on the CascadeStudio github page. But as I am using direct calls to opencascade.js, perhaps viewers of this repository can help: 

I am having difficulty creating a solid out of a shell. I wanted to use this functionality to create a socalled icosphere or geodesic dome. I used some code that I found on the internet to create these shapes. To create an icosphere with a certain subdivision level you have to start with an icosahedron. I created all faces using the standard Polygon() function of CascadeStudio. Then I sewed all faces to a shell. I confirmed that the shell is closed. Then I tried to create a solid out of the closed shell using  oc.BRepBuilderAPI_MakeSolid(). After this I checked that indeed a solid was created.  

The following scene shows my result. I get a shape that I can scale and translate. The scene contains three shapes, namely a box, a sphere and the icosphere. All three shapes are identified as solids (ShapeType()=2). A boolean operation between the box and the sphere works, but if you uncomment the last line in my code, containing a boolean of the icosphere with the resulting shape, the icosphere disappears. 

What am I doing wrong? 

 
![icosphere](https://user-images.githubusercontent.com/38007983/170941461-b5e7d81c-961d-4620-a58d-523bd5f66bb8.jpg)

```` javascript
// icospheres

function ProjectSphere(radius, vertex)
// function to project a vertex on to a sphere with radius "radius"
{
    let x = vertex[0];
    let y = vertex[1];
    let z = vertex[2];
    let currentRadius = Math.sqrt(Math.pow(x,2)+Math.pow(y,2)+Math.pow(z,2))
    let scale = radius/currentRadius;
    let scaledVertex = [scale*x , scale*y, scale*z]
    return  scaledVertex
}

function Icosahedron(radius)
{
    let golden = (1+Math.sqrt(5))/2
    let v = []
    // vertices determined by 4 rectangles
    v[0] = ProjectSphere(radius,[-1,  golden, 0]);
    v[1] = ProjectSphere(radius,[ 1,  golden, 0]);
    v[2] = ProjectSphere(radius,[-1, -golden, 0]);
    v[3] = ProjectSphere(radius,[ 1, -golden, 0]);

    v[4] = ProjectSphere(radius,[ 0, -1, golden]);
    v[5] = ProjectSphere(radius,[ 0,  1, golden]);
    v[6]= ProjectSphere(radius,[ 0, -1,-golden]);
    v[7] = ProjectSphere(radius,[ 0,  1,-golden]);

    v[8] = ProjectSphere(radius,[ golden,0, -1]);
    v[9] = ProjectSphere(radius,[ golden,0,  1]);
    v[10]= ProjectSphere(radius,[ -golden,0, -1]);
    v[11] = ProjectSphere(radius,[ -golden,0, 1]);

    // faces added so that they always have an edge in common
    // with the previous ones
    let f = []
    f[0] = [v[0],v[11],v[5]]
    f[1] = [v[0],v[5],v[1]]
    f[2] = [v[0],v[10],v[11]]
    f[3] = [v[0],v[7],v[10]]
    f[4] = [v[5],v[11],v[4]]
    f[5] = [v[4],v[9],v[5]]
    f[6] = [v[3],v[9],v[4]]
    f[7] = [v[3],v[8],v[9]]
    f[8] = [v[3],v[6],v[8]]
    f[9] = [v[3],v[2],v[6]]
    f[10] = [v[6],v[2],v[10]]
    f[11] = [v[10],v[7],v[6]]
    f[12] = [v[8],v[6],v[7]]
    f[13] = [v[0],v[1],v[7]]  
    f[14] = [v[1],v[5],v[9]]  
    f[15] = [v[11],v[10],v[2]] 
    f[16] = [v[7],v[1],v[8]]  
    f[17] = [v[3],v[4],v[2]]  
    f[18] = [v[2],v[4],v[11]] 
    f[19] = [v[9],v[8],v[1]]  

    // create polygons/faces out of the 20 * three vertices
    // and sew them together to a closed shell
    let icosphere = new oc.BRepBuilderAPI_Sewing()
    for(let i=0; i<=19 ;i++)
    {
        face = Polygon(f[i])
        icosphere.Add(face)
        sceneShapes.pop(face)  // to avoid that Polygon() puts faces in the scene
    }
    icosphere.Perform()
    let sewedsphere = icosphere.SewedShape()
    console.log(sewedsphere.ShapeType())  // check type 3 (shell) of the result
    console.log(sewedsphere.Closed())  // check if the shell is closed (true)
    
    // create a solid
    let result = new oc.BRepBuilderAPI_MakeSolid()
    result.Add(sewedsphere)
    let finalresult = result.Solid() 
    console.log(finalresult.ShapeType())
    return finalresult
}

// draw the isosphere
let test= Icosahedron(2.0);

test = Scale(50,test,false)
test = Translate([20,20,220],test,false)
console.log(test.ShapeType())

// draw a box
let box = Box(200,200,200)
console.log(box.ShapeType())

// draw a standard sphere
let studioSphere = Translate([50,30,20],Sphere(100))
console.log(studioSphere.ShapeType())

box = Difference(box,[studioSphere])
// let experiment = Difference(box,[test])

````

**Comment by sgenoud** [comment]
Hello,

I could make it [work with replicad](https://studio.replicad.xyz/share/https%3A%2F%2Fgist.githubusercontent.com%2Fsgenoud%2F2d9dfcc3bf89bc3b58168352284da595%2Fraw%2F9d21ba14548f91e982724976d20b702068bf5371%2Ficosahedron.js)! (you can click on the source link at the bottom to see the source).

I do not use `MakeSolid` but `ShapeFix_Solid().SolidFromShell(shell)`, as you can [see here](https://github.com/sgenoud/replicad/blob/main/packages/replicad/src/shapeHelpers.ts#L429). It is a bit more robust for weirder shapes! (it might also be related of the version of opencascade that is a bit older in cascade studio)

**Reply by raydeleu** [reply]
That is great, thanks for your effort! I will dive into your code to see what is happening and if I can translate what I have learned in CascadeStudio to Replicad. I will also look into the MakeThickSolid which I also do not get to work in CascadeStudio but seems to work fine in Replicad.
